Add tests to verify if discovery service is started from LabVIEW

jayaseelan-james opened the pull request:
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-labview/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?
Adds test to verify if discovery service is started from LabVIEW and includes it in the github workflows pipeline.

### Why should this Pull Request be merged?
Add automated tests in the PR pipeline to check if discovery service is running when starting a LabVIEW measurement service.

### What testing has been done?
1. Manually tested the code with and without discovery service running. 
2. Executed the `run_tests.vi` and confirmed that all the tests are passing.

Snapshots of the new VIs:

`DiscoveryServiceNotRunning_LaunchDiscoveryService_DiscoveryServiceStarted.vi`:
![image](https://github.com/ni/measurementlink-labview/assets/104884050/5221ab61-cc08-4350-9984-b585ab33e99a)

`Kill Discovery Service.vi`:
![image](https://github.com/ni/measurementlink-labview/assets/104884050/5bd8b4b5-8a29-436b-8213-4b519ae60891)

`Check If Discovery Service is Running.vi`:
![image](https://github.com/ni/measurementlink-labview/assets/104884050/a02103e7-7c00-456c-a98b-5593726e3e2b)

Jayaseelan-James commented:
NOTE: We have added the test to verify if discovery service is launched as the last test among all the tests. So, adding tests after this in `Tests.lst` containing a pre-requisite to have discovery service running will also fail if this particular test fail in the pipeline. 

Please add your thoughts on these lines as well.

dixonjoel commented:
> NOTE:
> 
> 1. I updated the Access Scope of Private Helpers folder from `Community` to `Public` to access the VIs outside the library to add tests.
> 2. I have added the test to verify if discovery service is launched as the last test among all the tests. So, adding tests after this in `Tests.lst` containing a pre-requisite to have discovery service running will also fail if this particular test fail in the pipeline.
> 
> Please add your thoughts on these lines as well.

I don't like making all of those helper VIs public. I think instead we should move only 'Ensure Discovery Service Started.vi' into the Public Helpers folder. Then we modify the test to call this VI after it calls kill discovery service.

As is, the test is really not validating that the DS is running, only that the key file exists and is locked. So the assert end of the test could actually verify that NationalInstruments.MeasurementLink.Discovery.exe is actually running. You could use something like this:
https://forums.ni.com/t5/Example-Code/How-to-get-a-list-of-System-Process-from-within-LabVIEW/ta-p/3522753

Jayaseelan-James commented:
> > NOTE:
> > 
> > 1. I updated the Access Scope of Private Helpers folder from `Community` to `Public` to access the VIs outside the library to add tests.
> > 2. I have added the test to verify if discovery service is launched as the last test among all the tests. So, adding tests after this in `Tests.lst` containing a pre-requisite to have discovery service running will also fail if this particular test fail in the pipeline.
> > 
> > Please add your thoughts on these lines as well.
> 
> I don't like making all of those helper VIs public. I think instead we should move only 'Ensure Discovery Service Started.vi' into the Public Helpers folder. Then we modify the test to call this VI after it calls kill discovery service.
> 
> As is, the test is really not validating that the DS is running, only that the key file exists and is locked. So the assert end of the test could actually verify that NationalInstruments.MeasurementLink.Discovery.exe is actually running. You could use something like this: https://forums.ni.com/t5/Example-Code/How-to-get-a-list-of-System-Process-from-within-LabVIEW/ta-p/3522753

The reason for changing the access scope of Private Helpers from `Community` to `Public` when adding tests, instead of moving the VIs to Public Helpers is:
- Since all this code is open-source to the users (once installed in vi.lib - user can do anything with this), I felt it won't make much difference if I change the access of "Private Helpers" folder instead of placing these specific VIs (required for the tests) in "Public Helpers". When I place the `Ensure Discovery Service Started.vi` & `Start Discovery Service.vi` VIs under Public Helpers, this might give our users the sense that these VIs are intended for public use which isn't the ideal case. Since these VIs are discovery service related, I am hoping users should not be using it. Hence placing it in Private Helpers folder is the right way.


Whereas, though the folder is not restricted with Community access, having the VIs within the Private Helpers would still mean that they're not supposed to be modified (from the Private helpers folder name).


Now, I would like to understand if there was any specific reason(s) to have the Private Helpers folder restricted with Community access.

Jayaseelan-James commented:
> As is, the test is really not validating that the DS is running, only that the key file exists and is locked. So the assert end of the test could actually verify that NationalInstruments.MeasurementLink.Discovery.exe is actually running. You could use something like this: https://forums.ni.com/t5/Example-Code/How-to-get-a-list-of-System-Process-from-within-LabVIEW/ta-p/3522753

Agreed, @dixonjoel. I'll update the assert condition to check if `NationalInstruments.MeasurementLink.DiscoveryService.exe` is present in the process list.

dixonjoel commented:
> Now, I would like to understand if there was any specific reason(s) to have the Private Helpers folder restricted with Community access.

https://labviewwiki.org/wiki/Access_scope#:~:text=Access%20scope%20defines%20the%20accessibility,scope%20to%20all%20contained%20items.

Community access is pretty similar to Private, except it allows the possibility that there could be 'friend' libraries that can call those VIs. Since we're not using this right now, I would be ok with making it Private. However, it's not good to expose the implementation details to customers by allowing them to call these Private VIs.

Ensure Discovery Service Started.vi is something that I think is reasonable to be a public API even though we chose not to expose it initially. I could envision scenarios where the user wanted to get the discovery service started.

As to the vi.lib, just putting it in vi.lib doesn't really mean it's open source or that users are expected to or allowed to change it. They would do that at their own peril and it's well understood that you shouldn't change contents of vi.lib. Being able to call something from vi.lib is a different question and the users should only be able to call things that make sense as a public API. Putting Ensure Discovery Service Started as a public VI is a bit of a compromise so we don't have to expose the whole Private Helpers. Another option might be to keep it in the Private Helpers folder but *only* make that VI public (it doesn't have to inherit the folder access scope I think).

Jayaseelan-James commented:
Thanks for the clarification, @dixonjoel. I went on with the approach to update the access scope of "Private Helpers" to `Not Specified` and exposing only the `Ensure Discovery Service Started.vi`, `Start Discovery Service.vi` and `Get Discovery Service Key File Path.vi` as Public.

Jayaseelan-James commented:
> So, adding tests after this in `Tests.lst` containing a pre-requisite to have discovery service running will also fail if this particular test fail in the pipeline.

@dixonjoel, could you please share your thoughts on the above quoted statement as well?

Jayaseelan-James commented:
@aepete, since you worked on creating tests for the other libraries, I would like to know your comments on this PR changes.

dixonjoel reviewed (approved):
Approved if the issue about Ensure Discovery Service Running.vi being public is addressed.

Jayaseelan-James commented on Source/Runtime/MeasurementLink Measurement Server/MeasurementLink Measurement Server.lvlib:
```diff
@@ -8,32 +8,73 @@
+		<Item Name="Encode to JSON String.vim" Type="VI" URL="../Helpers/Encode to JSON String.vim">
+			<Property Name="NI.LibItem.Scope" Type="Int">4</Property>
+		</Item>
 		<Item Name="Ensure Discovery Service Started.vi" Type="VI" URL="../Helpers/Ensure Discovery Service Started.vi"/>
```
Yes, `Ensure Discovery Service Started.vi` is not used in the test as it's just a wrapper over `Start Discovery Service.vi`. The reason for making it public is, since `Start Discovery Service.vi` is made public having it's wrapper `Private` seemed unnecessary.

dixonjoel replied:
I would lean towards not making things public unless required. However, I think this is a reasonable one to be part of the 'sort-of public' API. So I'll approve with this public.
